### PR TITLE
List exact regs killed by ByRefWriteBarrier on AMD64

### PR DESCRIPTION
**Pull request by EgorBo**
Seems that ARM64 is already fine as is and listing `r12,r15,ip0,ip1 (r17)` as trashed registers.

### CODE REVIEW

**Comment by EgorBo** [comment]
@dotnet/jit-contrib PTAL, [diffs on x64](https://dev.azure.com/dnceng-public/public/_build/results?buildId=585307&view=ms.vss-build-web.run-extensions-tab). Arm64 is fine as is. We might explore other helpers as well but I coudn't find more obvious cases (many helpers are written in C++ or C# so we can't really assume their registers).

**Review by jakobbotsch — COMMENTED** [review]
The comment on `compHelperCallKillSet` should be deleted -- I assume it's not correct anymore if this change just works as is. Also seems it would be good to run extended testing on this PR.

File: src/coreclr/jit/targetamd64.h
Comment (jkotas) [reply src/coreclr/jit/targetamd64.h:210]: There is no reason for the write barrier to trash different registers on regular CoreCLR vs. native AOT. It is easy to make it the same.

> Windows-CoreCLR-x64 does touches R10/R11 too, but it push/pop them,

It is only on debug-only GCStress path.

**Comment by kunalspathak** [comment]
> Right, in fact x86, arm and arm64 already define exact regs, it's only x64 that does not

I have fixed the PR title accordingly.

File: src/coreclr/jit/codegencommon.cpp
Comment (kunalspathak) [inline src/coreclr/jit/codegencommon.cpp:598]: We haven't done the "exact regs killed" for other helpers, so probably this comment should stay until we do it?

File: src/coreclr/jit/targetamd64.h
Comment (EgorBo) [reply src/coreclr/jit/targetamd64.h]: Yep, but since 3 out of 4 x64 platforms need them - I decided to include it conservatively into Windows as well. So:

1. `Windows-x64-CoreCLR`:
2. `Windows-x64-Naot`: needs R10, R11
3. `Unix-x64-CoreCLR`: needs R10, R11
4. `Unix-x64-Naot`: needs R10, R11

Added a comment. This costs few missing diff improvements on `Windows-x64-CoreCLR` so I decided to leave as is, the proper fix won't be simple as `#ifdef` since IsNaot + IsWindows/IsUnix checks aren't compile-time.

File: src/coreclr/jit/codegencommon.cpp
Comment (EgorBo) [reply src/coreclr/jit/codegencommon.cpp:598]: Added a comment

**Comment by BruceForstall** [comment]
This is https://github.com/dotnet/runtime/issues/67707

File: src/coreclr/jit/targetamd64.h
Comment (BruceForstall) [reply src/coreclr/jit/targetamd64.h]: This is not someplace we should be conservative. We should be precise. If Windows-x64-CoreCLR doesn't need R10/R11 in Release, then it shouldn't use them (or at least shouldn't trash them), and the JIT should be free to use them.

You can change the `#define` to call a function that checks the OS/ABI.

Is there any reason why the debug-only code path on Linux can't push/pop R10/R11 as well? (Maybe stack walking prevents it?)

**Review by BruceForstall — CHANGES_REQUESTED** [review]
I don't think we should be conservative on any platform: we should be precise.

File: src/coreclr/jit/targetamd64.h
Comment (jkotas) [reply src/coreclr/jit/targetamd64.h]: There is no reason for the reg alloc requirements for this helper to differ between Windows/Unix or JIT/AOT. The assembly code for the helper can be exactly the same for all variants. We ended up with different tweaks applied to different copies of the assembly code over time and that's why the register requirements differ.

Comment (EgorBo) [reply src/coreclr/jit/targetamd64.h]: > There is no reason for the reg alloc requirements for this helper to differ between Windows/Unix or JIT/AOT. The assembly code for the helper can be exactly the same for all variants. We ended up with different tweaks applied to different copies of the assembly code over time and that's why the register requirements differ.

@jkotas from a quick look, one issue that on windows (.asm) we can do:
```asm
cmp     rdi, [g_lowest_address]
```
while on Linux (.s) we have to do:
```asm
PREPARE_EXTERNAL_VAR g_lowest_address, rax
cmp     rdi, [rax]
```

Comment (jkotas) [reply src/coreclr/jit/targetamd64.h]: We should be able to the same on Linux too. We do that for native AOT, but not for regular CoreCLR for some reason: https://github.com/dotnet/runtime/blob/8330db998659c4e6410aba370b37e4304a517a2b/src/coreclr/nativeaot/Runtime/amd64/WriteBarriers.S#L181

Comment (EgorBo) [reply src/coreclr/jit/targetamd64.h]: Copy-pasted the CoreCLR-x64-windows version for Linux. Let's see..

Comment (EgorBo) [reply src/coreclr/jit/targetamd64.h]: ah, I should've tested it locally first, forgot about comments // and the syntax for labels

File: src/coreclr/jit/codegencommon.cpp
Comment (jkotas) [inline src/coreclr/jit/codegencommon.cpp:620]: There is no reason for native AOT to be different.

**Review by jkotas — CHANGES_REQUESTED** [review]
There is no reason for native AOT to be different.

File: src/coreclr/jit/codegencommon.cpp
Comment (EgorBo) [reply src/coreclr/jit/codegencommon.cpp:620]: Yeah it's not done yet, I am waiting for a green ci to make sure I did the unix version correctly. Do I need to copy the `WRITE_BARRIER_CHECK` pieces to NAOT version? it currently doesn't have it

**Comment by EgorBo** [comment]
@jkotas an extra reg (r11) might allow us to save an indirect load for when `FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP` is defined. but I am not sure it's worth it or not.

image

**Comment by jkotas** [comment]
> an extra reg (r11) might allow us to save an indirect load for when FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP is defined. but I am not sure it's worth it or not.

This is rare path. The static is zero most of the time. The static is non-zero only when the GC is running in the background and enabled write watch to keep track of the pages that we written to and needs to be revisited.

I think the CoreCLR shape is fine.

File: src/coreclr/vm/amd64/jithelpers_fast.S
Comment (jkotas) [inline src/coreclr/vm/amd64/jithelpers_fast.S:116]: This ifdef does not make sense. The code under the ifdef has nothing to do with software write watch.

I think you want to drop the comment that RAX is trashed only when `FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP` is defined. Instead, just say the RAX is trashed, end of story.

**Comment by EgorBo** [comment]
Diff between CoreCLR-Windows-x64 (I didn't change it) vs CoreCLR-Linux-x64 (I copied the Windows one and adjusted syntax): https://www.diffchecker.com/Sg9tWKl2/

**Comment by EgorBo** [comment]
@jkotas @BruceForstall can you take a look now? CI passed before my most recent change in comments. New [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=593705&view=ms.vss-build-web.run-extensions-tab) (a bit better)

**Comment by jkotas** [comment]
> I presume it's possible to do some asm code-sharing between NAOT and CoreCLR but if you don't mind I'd love to check that separately

Yes, there is a lot of sharing to be done between NAOT and CoreCLR in general. No need to do it in this PR.

File: src/coreclr/vm/amd64/jithelpers_fast.S
Comment (jkotas) [inline src/coreclr/vm/amd64/jithelpers_fast.S:114]:
```suggestion
```

File: src/coreclr/vm/amd64/jithelpers_fast.S
Comment (jkotas) [inline src/coreclr/vm/amd64/jithelpers_fast.S:163]:
```suggestion
```

**Review by jkotas — APPROVED** [review]
LGTM modulo comments

**Review by BruceForstall — APPROVED** [review]
You should probably run gcstress pipelines

**Pull request merged**
